compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeDefBuilderHelper.java

```diff
@@ -299,4 +307,72 @@ public static String getPackageAlias(SymbolEnv env, String compUnitName, Package
+                                                BStructureType origStructureType, Location pos,
+                                                SymbolEnv env, PackageID pkgID, Set<BType> unresolvedTypes,
+                                                long flag, boolean isImmutable, boolean isConstant) {
+        BTypeSymbol immutableStructureSymbol = structureType.tsymbol;
+        LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+        for (BField origField : origStructureType.fields.values()) {
+            BType fieldType;
+            if (isImmutable) {
+                fieldType = ImmutableTypeCloner.getImmutableType(pos, types, origField.type, env,
+                        env.enclPkg.packageID, env.scope.owner, symTable, anonymousModelHelper, names, unresolvedTypes);
+            } else {
+                fieldType = origField.type;
+            }
+
+            Name origFieldName = origField.name;
+            BVarSymbol immutableFieldSymbol;
+            if (fieldType.tag == TypeTags.INVOKABLE && fieldType.tsymbol != null) {
+                immutableFieldSymbol = new BInvokableSymbol(origField.symbol.tag, origField.symbol.flags | flag,
+                        origFieldName, pkgID, fieldType,
+                        immutableStructureSymbol, origField.symbol.pos, SOURCE);
+                BInvokableTypeSymbol tsymbol = (BInvokableTypeSymbol) fieldType.tsymbol;
+                BInvokableSymbol invokableSymbol = (BInvokableSymbol) immutableFieldSymbol;
+                invokableSymbol.params = tsymbol.params == null ? null : new ArrayList<>(tsymbol.params);
+                invokableSymbol.restParam = tsymbol.restParam;
+                invokableSymbol.retType = tsymbol.returnType;
+                invokableSymbol.flags = tsymbol.flags;
+            } else if (fieldType == symTable.semanticError) {
+                // Can only happen for records.
+                immutableFieldSymbol = new BVarSymbol(origField.symbol.flags | flag | Flags.OPTIONAL,
+                        origFieldName, pkgID, symTable.neverType,
+                        immutableStructureSymbol, origField.symbol.pos, SOURCE);
+            } else {
+                immutableFieldSymbol = new BVarSymbol(origField.symbol.flags | flag, origFieldName, pkgID,
+                        fieldType, immutableStructureSymbol,
+                        origField.symbol.pos, SOURCE);
+            }
+            String nameString = origFieldName.value;
+            fields.put(nameString, new BField(origFieldName, null, immutableFieldSymbol));
+            immutableStructureSymbol.scope.define(origFieldName, immutableFieldSymbol);
+        }
+        structureType.fields = fields;
+
+        if (origStructureType.tag == TypeTags.OBJECT) {
+            return;
+        }
+
+        if (isConstant) {
+            ((BLangRecordTypeNode) structureTypeNode).sealed = true;
+            ((BRecordType) structureType).restFieldType = new BNoType(TypeTags.NONE);
+        }
```

Comment:
This doesn't seem like something that should be done here. Can't we do do this where it is called instead?

Reply:
Will update. Thanks.